src/reader.rs

```diff
@@ -438,3 +438,90 @@ fn fast_find(needle: &[u8], haystack: &[u8]) -> Option<usize> {
     #[cfg(not(feature = "jetscii"))]
     haystack.iter().position(|b| needle.contains(b))
 }
+
+/// A Reader to wrap other [`Reader`]s with to get spans.
+///
+/// This reader saves the current byte offset in the input stream in [`Self::position`].
+///
+/// Example:
+///
+/// ```rust
+/// use std::fmt::Write;
+/// use html5gum::{DefaultEmitter, SpanReader, Token, Tokenizer};
+///
+/// let tokenizer = Tokenizer::new_with_emitter(
+///     SpanReader::new("<title>hello world</title>"),
+///     DefaultEmitter::<_, usize>::new_with_span(),
+/// );
+///
+/// let mut spans = Vec::new();
+///
+/// for token in tokenizer {
+///     let token = token.unwrap();
+///
+///     match token {
+///         Token::StartTag(tag) => {
+///             spans.push((format!("start: {}", String::from_utf8_lossy(&tag.name)), tag.span.start));
+///         }
+///         Token::String(string) => {
+///             spans.push((format!("string: {}", String::from_utf8_lossy(&string)), string.span.start));
+///         }
+///         Token::EndTag(tag) => {
+///             spans.push((format!("end: {}", String::from_utf8_lossy(&tag.name)), tag.span.start));
+///         }
+///         _ => panic!("unexpected input"),
+///     }
+///
+/// }
+///
+/// assert_eq!(spans, vec![
+///     ("start: title".to_string(), 0),
+///     ("string: hello world".to_string(), 7),
+///     ("end: title".to_string(), 18),
+/// ]);
+/// ```
+#[derive(Debug)]
+pub struct SpanReader<R> {
+    /// The inner reader.
+    pub reader: R,
+    /// The current byte offset in the input stream.
+    pub position: usize,
+}
+
+impl<R: Reader> SpanReader<R> {
+    /// Construct a new [`SpanReader`] from any type that implements [`Readable`] (including other
+    /// [`Reader`]s!).
+    ///
+    /// `input` can be [`&String`][String], [`&str`][str], [`Vec<u8>`], [`&[u8]`][slice] or any
+    /// other [`Reader`] at the moment, as those are the types for which [`crate::Readable`] is
+    /// implemented, but you can implement that trait on your own types.
+    pub fn new<'a, S: Readable<'a, Reader = R>>(input: S) -> Self {
+        Self {
+            reader: input.to_reader(),
+            position: 0,
+        }
+    }
+}
+
+impl<R: Reader> Reader for SpanReader<R> {
+    type Error = R::Error;
+
+    fn read_byte(&mut self) -> Result<Option<u8>, Self::Error> {
+        match self.reader.read_byte()? {
+            Some(v) => {
+                self.position += 1;
+                Ok(Some(v))
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn try_read_string(&mut self, s: &[u8], case_sensitive: bool) -> Result<bool, Self::Error> {
```

Comment:
since `read_until` is not implemented, the `SpanReader` will not perform well for large contiguous blocks of text such as `<p>aaaaa.... aaaaa</p>`

I wonder if instead of having a span reader that tracks its position, changes to the position could be emitted as "events" similar to start tags, etc:

```
trait Emitter {
    fn advance_position(&mut self, offset: usize) {}
}
```

if the method is empty, then the compiler should be able to optimize away all overhead